include/dlaf/eigensolver/tridiag_solver/mc.h

```diff
@@ -0,0 +1,241 @@
+  static void call(Matrix<T, Device::CPU>& mat_trd, Matrix<T, Device::CPU>& d,
+                   Matrix<T, Device::CPU>& mat_ev);
+};
+
+/// Splits [i_begin, i_end] in the middle and waits for all splits on [i_begin, i_middle] and [i_middle +
+/// 1, i_end] before saving the triad <i_begin, i_middle, i_end> into `indices`.
+///
+/// The recursive calls span a binary tree which is traversed in depth first left-right-root order. That
+/// is also the order of triads in `indices`.
+///
+/// Note: the intervals are all closed!
+///
+inline void splitIntervalInTheMiddleRecursively(
+    SizeType i_begin, SizeType i_end, std::vector<std::tuple<SizeType, SizeType, SizeType>>& indices) {
+  if (i_begin == i_end)
+    return;
+  SizeType i_middle = (i_begin + i_end) / 2;
+  splitIntervalInTheMiddleRecursively(i_begin, i_middle, indices);
+  splitIntervalInTheMiddleRecursively(i_middle + 1, i_end, indices);
+  indices.emplace_back(i_begin, i_middle, i_end);
+}
+
+/// Generates an array of triad indices. Each triad is composed of begin <= middle < end indices and
+/// represents two intervals [begin, middle] and [middle + 1, end]. The two intervals are the subprobelms
+/// that have to be merged to arrive at [begin, end].
+///
+/// Note: the intervals are all closed!
+///
+inline std::vector<std::tuple<SizeType, SizeType, SizeType>> generateSubproblemIndices(SizeType n) {
+  DLAF_ASSERT(n > 0, n);
+  std::vector<std::tuple<SizeType, SizeType, SizeType>> indices;
+  indices.reserve(to_sizet(n));
+  splitIntervalInTheMiddleRecursively(0, n - 1, indices);
+  return indices;
+}
+
+// Cuppen's decompostion
+//
+// Substracts the offdiagonal element at the split from the top and bottom diagonal elements and returns
+// the offdiagonal element. The split is between the last row of the top tile and the first row of the
+// bottom tile.
+//
+template <class T>
+T cuppensTileDecomposition(const matrix::Tile<T, Device::CPU>& top,
+                           const matrix::Tile<T, Device::CPU>& bottom) {
+  (void) top;
+  (void) bottom;
+
+  T offdiag_val = top(TileElementIndex{top.size().rows() - 1, 1});
+  T& top_diag_val = top(TileElementIndex{top.size().rows() - 1, 0});
+  T& bottom_diag_val = bottom(TileElementIndex{0, 0});
+
+  top_diag_val -= offdiag_val;
+  bottom_diag_val -= offdiag_val;
+  return offdiag_val;
+}
+
+DLAF_MAKE_CALLABLE_OBJECT(cuppensTileDecomposition);
+
+template <class T>
+std::vector<pika::shared_future<T>> cuppensDecomposition(Matrix<T, Device::CPU>& mat_trd) {
+  SizeType i_end = mat_trd.distribution().nrTiles().rows() - 1;
+  std::vector<pika::shared_future<T>> offdiag_vals;
+  offdiag_vals.reserve(to_sizet(i_end));
+
+  for (SizeType i_split = 0; i_split < i_end; ++i_split) {
+    // Cuppen's tridiagonal decomposition
+    offdiag_vals[to_sizet(i_split)] =
+        pika::dataflow(pika::unwrapping(cuppensTileDecomposition_o), mat_trd(LocalTileIndex(i_split, 0)),
+                       mat_trd(LocalTileIndex(i_split + 1, 0)));
```

Comment:
```suggestion
        when_all(mat_trd.readwrite_sender(LocalTileIndex(i_split, 0)),
                       mat_trd.readwrite_sender(LocalTileIndex(i_split + 1, 0))) | transform(cuppensTileDecomposition_o) | make_future();
```
modulo namespaces/typos/unclosed parentheses...